File: backend/schemas/trace_schema.py
"""
TraceFlow Trace Schema — v1.0 (FROZEN)
=======================================

This is the single canonical schema used across every phase:
- Phase 4 (current): mock traces served from mock_traces.json.
- Phase 5:           Java parser output MUST conform to this schema.
- Phase 6:           Execution trace generator MUST produce Step objects
                     matching this shape.
- Phase 7-8:         Frontend variable & output UI consume `variables`,
                     `output`, `changes` fields directly — no adapter needed.
- Phase 9:           LLM explanation writes to `explanation` field only;
                     rest of the object is unchanged.

Rule of thumb: DO NOT add required fields here without bumping schema_version
and updating the frontend consumer components (VariableCard, ExecutionPanel,
OutputConsole, AIExplanation).

Optional fields (`kind`, `label`, `condition`, `condition_result`) are UI
hints; the frontend degrades gracefully when they are absent.
"""

from __future__ import annotations
from typing import Any, Dict, List, Optional, Literal
from pydantic import BaseModel, Field


SCHEMA_VERSION: str = "1.0"


class Step(BaseModel):
    """One frame in the execution trace. Order in trace.steps == chronological."""

    step: int = Field(..., ge=1, description="1-indexed monotonically increasing step number")
    line: int = Field(..., ge=1, description="1-indexed source line currently executing")
    variables: Dict[str, Any] = Field(
        default_factory=dict,
        description="Full snapshot of variable name -> value AFTER this step",
    )
    output: List[str] = Field(
        default_factory=list,
        description="Full print buffer as a list of strings AFTER this step",
    )
    changes: List[str] = Field(
        default_factory=list,
        description="Human-readable list of what changed on this step (e.g. 'sum changed from 0 to 1')",
    )
    explanation: str = Field(
        default="",
        description="Short (<=3 sentence) explanation. Mocked in Phase 4, LLM-generated in Phase 9",
    )

    # ----- optional UI hints (safe to omit) -----
    kind: Optional[
        Literal[
            "declare", "assign", "condition", "loop-init", "loop-step", "print", "call", "return"
        ]
    ] = None
    label: Optional[str] = Field(default=None, description="Short mono label shown in UI")
    condition: Optional[str] = None
    condition_result: Optional[bool] = None


class Trace(BaseModel):
    """A full program trace."""

    id: str
    name: str
    description: str = ""
    concept: Optional[str] = None
    code: str
    steps: List[Step]


class TraceCatalogEntry(BaseModel):
    id: str
    name: str
    description: str = ""

File: frontend/src/schemas/traceSchema.js
/**
 * TraceFlow Trace Schema — v1.0 (FROZEN)
 * =======================================
 *
 * Mirrors backend/schemas/trace_schema.py. Every step served by the API
 * (mock in Phase 4, parser output in Phase 5+, LLM-explained in Phase 9)
 * MUST conform to this shape. Optional fields are UI hints only.
 *
 * @typedef {Object} Step
 * @property {number} step        1-indexed chronological step id
 * @property {number} line        1-indexed source line executing
 * @property {Object.<string, *>} variables  Full variable snapshot AFTER this step
 * @property {string[]} output    Full print buffer AFTER this step
 * @property {string[]} changes   Human-readable list of what changed here
 * @property {string} explanation Short (<=3 sentence) explanation
 *
 * -- optional UI hints (may be absent) --
 * @property {("declare"|"assign"|"condition"|"loop-init"|"loop-step"|"print"|"call"|"return")=} kind
 * @property {string=} label
 * @property {string=} condition
 * @property {boolean=} condition_result
 *
 * @typedef {Object} Trace
 * @property {string} id
 * @property {string} name
 * @property {string} description
 * @property {string=} concept
 * @property {string} code
 * @property {Step[]} steps
 */

export const SCHEMA_VERSION = "1.0";

/** Required fields on a Step — used only by the Trace Inspector for validation display. */
export const STEP_REQUIRED_FIELDS = [
  "step",
  "line",
  "variables",
  "output",
  "changes",
  "explanation",
];

/**
 * Return the set of variable names whose value differs between prev and current
 * step. Derived from the canonical `variables` snapshot — DO NOT rely on
 * optional trace fields for this.
 * @param {Step|null} current
 * @param {Step|null} prev
 * @returns {Set<string>}
 */
export function diffChangedVars(current, prev) {
  const changed = new Set();
  if (!current) return changed;
  const cur = current.variables || {};
  const p = prev?.variables || {};
  for (const key of Object.keys(cur)) {
    // Compare by JSON to handle numbers/booleans/strings uniformly
    if (JSON.stringify(p[key]) !== JSON.stringify(cur[key])) {
      changed.add(key);
    }
  }
  return changed;
}

/** Lightweight schema validator used by the Trace Inspector to surface issues early. */
export function validateStep(step) {
  const problems = [];
  if (!step || typeof step !== "object") {
    return ["step is not an object"];
  }
  for (const key of STEP_REQUIRED_FIELDS) {
    if (!(key in step)) problems.push(`missing required field: ${key}`);
  }
  if ("variables" in step && (step.variables === null || typeof step.variables !== "object")) {
    problems.push("variables must be an object");
  }
  if ("output" in step && !Array.isArray(step.output)) {
    problems.push("output must be an array of strings");
  }
  if ("changes" in step && !Array.isArray(step.changes)) {
    problems.push("changes must be an array of strings");
  }
  return problems;
}
